### PR TITLE
Add SimpleSequence as preferred alias for SupportsLenAndGetItem

### PR DESCRIPTION
**Pull request by srittau**
This protocol represents a sequence as defined in the Python
documentation, as opposed to collections.abc.Sequence.

### CODE REVIEW

**Comment by srittau** [comment]
Maybe we can find a better name? But it should have `Sequence` in there.

**Comment by Akuli** [comment]
I think the name should be super obvious, so that you immediately know what it does when you see it somewhere in the code, such as `SupportsLenAndGetItem` :)

**Comment by srittau** [comment]
There already is a concept in Python that expresses this, "sequence". We should use this name to make it clear that this protocol actually reflect that concept.

**Comment by Akuli** [comment]
What about functions that only require the sequence to have a `__getitem__` method, such as `iter()`? There is no one single type that describes "what you actually want to use when it's tempting to use `Sequence`", as that may or may not require `__len__` to exist, so I don't think we should make a protocol whose name suggest that it is "what you actually want to use when it's tempting to use `Sequence`".

**Comment by Akuli** [comment]
Actually nevermind. A type like "what docs describe as `Sequence`" is fine, because it is very uncommon to implement `__getitem__` without `__len__`, so I don't think this will be a problem in practice. And in many cases, `SupportsLenAndGetItem` is an improvement over `Sequence`.

As a side note, I generally consider `Sequence` a bit of a code smell, because often the reality is "the code for loops over it" or "the code checks for lists and tuples with `isinstance`".

File: stdlib/_typeshed/__init__.pyi [hunk 1]
```diff
@@ -46,6 +46,15 @@ class SupportsDivMod(Protocol[_T_contra, _T_co]):
...
 
+# This represents a sequence as defined in the Python documentation:
+# https://docs.python.org/3/glossary.html#term-sequence
+# collections.abc.Sequence and typing.Sequence are not protocols and
+# have additional fields that are often unnecessary.
+class SimpleSequence(Protocol[_T_co]):
+    def __len__(self) -> int: ...
+    def __getitem__(self, __k: int) -> _T_co: ...
+
+# obsolete, use SimpleSequence instead
```
Comment (Akuli) [inline stdlib/_typeshed/__init__.pyi:57]: Why not just delete it? It was not marked with `# stable`.

Comment (Akuli) [inline stdlib/_typeshed/__init__.pyi:57]: Why not just delete it? It was not marked with `# stable`.

**Review by Akuli — COMMENTED** [review]
I intended to include this in my other review, but I think I clicked the wrong button.

**Edit:** Now my first review comment was lost somehow, I'll just write it here instead of trying to add another review and losing something again :) The new name should say "it is sequence the concept, not the class" or "it has `__len__` and `__getitem__`". So something like `SequenceAccordingToDocs`, but that isn't a great name either.

File: stdlib/_typeshed/__init__.pyi (hunk 1 above)
Comment (hauntsaninja) [reply stdlib/_typeshed/__init__.pyi:57]: I guess if it's used in third party stubs (there seems to be the one single case), under some circumstances you could run into silent loss of type checking. This already ends up happening though, e.g. see #5786 and #5751, so I'd also be inclined to just delete it.

If we don't delete it, we should alias here, though, and still update the `annoy` stubs

**Review by hauntsaninja — COMMENTED** [review]
I don't feel strongly about this name vs SupportsLenAndGetItem. Typing users I interact with still get confused about just Sequence vs List :-)

File: stdlib/_typeshed/__init__.pyi (hunk 1 above)
Comment (srittau) [reply stdlib/_typeshed/__init__.pyi:57]: The annoy stubs were my main concern. If we update them, they won't work, until type checkers have been updated with the latest typeshed. I'd suggest keeping it for a few months.

**Comment by srittau** [comment]
> **Edit:** Now my first review comment was lost somehow, I'll just write it here instead of trying to add another review and losing something again :) The new name should say "it is sequence the concept, not the class" or "it has `__len__` and `__getitem__`". So something like `SequenceAccordingToDocs`, but that isn't a great name either.

`PythonSequence`, `GlossarySequence`, `NotThatOtherSequence`, `RealSequence`

**Comment by Akuli** [comment]
As a maintainer, I would love it to be `NotThatOtherSequence`, but that isn't exactly very contributor friendly because most other people don't know the context.

**Comment by Akuli** [comment]
How about `GetItemAndLenSequence`? It seems to satisfy everyone's needs, as it very clearly says what an object must have to satisfy the protocol, and the name also hints that it is an alternative to the usual `Sequence`. I think it would be an improvement over `SupportsLenAndGetItem`, because there are mappings that also support `__len__` and `__getitem__`, but the protocol isn't intended to be used with those.

**Comment by BvB93** [comment]
Yet another option would bo something like `PySequence`; this is the name used in the [C-API](https://docs.python.org/3/c-api/sequence.html) to refer to this type of protocol.

**Comment by JelleZijlstra** [comment]
I don't like `PySequence`; if a user sees a `Sequence` and a `PySequence`, how are they supposed to know which is which?

I'm OK with most other options presented here, but I prefer `SimpleSequence` for brevity or `SupportsGetItemAndLen` (the current version) for explicitness.

**Comment by hauntsaninja** [comment]
I don't think we have consensus here, so closing as per https://github.com/python/typeshed/pull/6563#issuecomment-991263229

**Pull request closed without merging**